packages/query-persist-client-core/src/createPersister.ts

```diff
@@ -0,0 +1,147 @@
+import type {
+  QueryClient,
+  QueryFunctionContext,
+  QueryKey,
+  QueryState,
+} from '@tanstack/query-core'
+
+export interface PersistedQuery {
+  buster: string
+  timestamp: number
+  queryHash: string
+  queryKey: QueryKey
+  state: QueryState
+}
+
+export interface AsyncStorage {
+  getItem: (key: string) => Promise<string | undefined | null>
+  setItem: (key: string, value: string) => Promise<void>
+  removeItem: (key: string) => Promise<void>
+}
+
+export interface StoragePersisterOptions<QC extends QueryClient> {
+  /**
+   * Query Client instance
+   */
+  queryClient: QC
+  /** The storage client used for setting and retrieving items from cache.
+   * For SSR pass in `undefined`.
+   */
+  storage: AsyncStorage | Storage | undefined | null
+  /**
+   * How to serialize the data to storage.
+   * @default `JSON.stringify`
+   */
+  serialize?: (client: PersistedQuery) => string
+  /**
+   * How to deserialize the data from storage.
+   * @default `JSON.parse`
+   */
+  deserialize?: (cachedString: string) => PersistedQuery
+  /**
+   * A unique string that can be used to forcefully invalidate existing caches,
+   * if they do not share the same buster string
+   */
+  buster?: string
+  /**
+   * The max-allowed age of the cache in milliseconds.
+   * If a persisted cache is found that is older than this
+   * time, it will be discarded
+   */
+  maxAge?: number
+}
+
+/**
+ * Warning: experimental feature.
+ * This utility function enables fine-grained query persistance.
+ * Simple add it as a `persister` parameter to `useQuery` or `defaultOptions` on `queryClient`.
+ * 
+ * ```
+ * useQuery({
+     queryKey: ['myKey'],
+     queryFn: fetcher,
+     persister: createPersister({
+       storage: localStorage,
+       queryClient,
+     }),
+   })
+   ```
+ */
+export function createPersister<T, QC extends QueryClient>({
+  queryClient,
+  storage,
+  buster = '',
+  maxAge = 1000 * 60 * 60 * 24,
+  serialize = JSON.stringify,
+  deserialize = JSON.parse,
+}: StoragePersisterOptions<QC>) {
+  return async (
+    queryFn: (context: QueryFunctionContext) => T | Promise<T>,
+    context: QueryFunctionContext,
+  ) => {
+    const queryHash = hashKey(context.queryKey)
+    const queryState = queryClient.getQueryState(context.queryKey)
+
+    // Try to restore only if we do not have any data in the cache and we have persister defined
+    if (!queryState?.data && storage != null) {
+      const storedData = await storage.getItem(queryHash)
+      if (storedData) {
+        const persistedQuery = deserialize(storedData)
+
+        if (persistedQuery.timestamp) {
+          const expired = Date.now() - persistedQuery.timestamp > maxAge
+          const busted = persistedQuery.buster !== buster
+          if (expired || busted) {
+            await storage.removeItem(queryHash)
+          } else {
+            // Just after restoring we want to get fresh data from the server
+            // Maybe add an option for this?
```

Comment:
I think we'd want to compare the timestamp of the persisted Query with the current date by using the `timeUntilStale` util and the `staleTime` of the Query:

https://github.com/TanStack/query/blob/13b17ee45de3299dd3d3a2dfc3d23ad2f024882a/packages/query-core/src/utils.ts#L88-L90

with that, we would know if we should refetch or not, based on weather the data we restore from storage is considered fresh or not ...

Question is really where to get the `staleTime` from, because that's an observer level setting, and we don't really have access to the QueryObserver from here 🤔 . Or do we ?

Reply:
I think we can:

- get the `query` from the `QueryClient` (by finding it via the `queryKey`)
- then, loop through all observers and take the smallest staleTime we find

similar to how we do it when there's a window focus refetch:

https://github.com/TanStack/query/blob/9c721e548f500be924da8ef6da041a919f513140/packages/query-core/src/query.ts#L273-L278

- if we have no observer, because we wind up executing the queryFunction via prefetching or so, there won't be a refetch.

I think this would be fine?

Reply:
I managed to solve it differently

Reply:
nice, I like it 🙌